File: RB/ts/RiotBase/Db/Core/DbTable.ts
import SortedArray from "../../System/SortedArray";
import DbItem      from "./DbItem";

//
export default
class DbTable {
    constructor() {
        this.data = new SortedArray<DbItem>(); // Inode -> DbItem. SortedArray is certainly not a perfect choice
    }

    insert( db_item ) : DbItem {
        this.data.set( db_item.inode, db_item );
        return db_item;
    }

    find( inode ) : DbItem {
        return this.data.get( inode );
    }

    data: SortedArray<DbItem> // Inode -> DbItem. SortedArray is certainly not a perfect choice
}

File: RB/cpp/RiotBase/OldDb/Core/DbTable.cpp
#include "../../Variable/Core/Methods.h"
#include "DbTable.h"

namespace RiotBase {

DbTable::~DbTable() {
    clear();
}

DbItem *DbTable::find( Inode inode ) {
    Map::iterator iter = data.find( inode, Cmp() );
    return iter != data.end() ? &*iter : 0;
}

DbItem *DbTable::insert( DbItem *item ) {
    data.insert( *item );
    return item;
}

void DbTable::clear() {
    Vec<DbItem *> items;
    for( DbItem &item : data )
        items << &item;
    data.clear();
    for( DbItem *item : items )
        dec_ref( item );

}

} // namespace RiotBase
